tracing/tracing.go

```diff
@@ -0,0 +1,59 @@
+// Package tracing handles opentracing support for skipper
+package tracing
+
+import (
+	"errors"
+	"strings"
+
+	instana "github.com/instana/golang-sensor"
+	lightstep "github.com/lightstep/lightstep-tracer-go"
+	ot "github.com/opentracing/opentracing-go"
+)
+
+var (
+	// ErrUnsupportedTracer is returned when an unsupported opentracing
+	// implementation was requested as tracer
+	ErrUnsupportedTracer error = errors.New("invalid argument, not a supported tracer")
+	// ErrMissingArguments is returned when an empty list is passed to Init()
+	ErrMissingArguments error = errors.New("no arguments passed")
+)
+
+// Init sets up opentracing
+func Init(opts []string) error {
+	if len(opts) == 0 {
+		return ErrMissingArguments
+	}
+	var impl string
+	impl, opts = opts[0], opts[1:]
+
+	var tracer ot.Tracer
+	switch impl {
+	case "noop":
+		tracer = &ot.NoopTracer{}
+
+	case "instana":
+		tracer = instana.NewTracerWithOptions(&instana.Options{
+			Service:  "skipper", // FIXME
+			LogLevel: instana.Error,
+		})
+
+	case "lightstep":
+		var token string
+		for _, o := range opts {
+			if strings.HasPrefix(o, "token=") {
+				token = o[6:]
+			}
+		}
+		tracer = lightstep.NewTracer(lightstep.Options{
+			AccessToken: token,
+			Tags: map[string]interface{}{
+				lightstep.ComponentNameKey: "skipper", // FIXME
+			},
+		})
+
+	default:
+		return ErrUnsupportedTracer
+	}
+	ot.SetGlobalTracer(tracer)
```

Comment:
On the long run, I have always regretted later when I chose to use global, shared objects. Based on this: https://github.com/opentracing/opentracing-go/blob/master/globaltracer.go , using the globalTracer of this libe doesn't bring much value, and it's basically just provides a global access to the same shared object. This is similar to the current log usage that I already regretted. Just a few examples that come to my mind:

- mocking it for some testing becomes harder, sometimes even resulting in false positives
- replacing the implementation with a different compatible one becomes harder
- someone with 'weird' use cases cannot so freely reuse the library, e.g. when one would want to run multiple skipper proxies in the same process

Maybe, these examples are too specific and we just should be pragmatic, like now with the logs, and if everybody's fine with the global shared object, then i'm just fine with it, too. But I'd still vote for explicitly passing the tracing to the objects that use it. E.g. the main initializes the tracer, passes it to the proxy object, the proxy object passes it to the context for every request, the filters get it from the context if they want to use it. Would this be too much more tedious than just using the package level var?

Opinions?

Reply:
I agree! For logging it can be convenient to not have to pass the logger around every time you want to log something. But for this use case I think it would actually be more convenient/clear with a tracer field on the Proxy.

Reply:
Fine with me (and it's an easy change ;-))